### PR TITLE
Fix issue 19904

### PR DESCRIPTION
**Pull request by raghavanone**
Fixes issue #19904

### CODE REVIEW

**Comment by raghavanone** [comment]
@sgugger  Please tag the right persons for reviewing this PR .

**Review by sgugger — APPROVED** [review]
Thanks for the fix @amyeroberts or @Rocketknight1 could you double-check this looks good?

File: src/transformers/models/clip/modeling_clip.py [hunk 1]
```diff
@@ -62,13 +62,14 @@ def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int]
 
 # contrastive loss function, adapted from
 # https://sachinruk.github.io/blog/pytorch/pytorch%20lightning/loss%20function/gpu/2021/03/07/CLIP.html
-def contrastive_loss(logits: torch.Tensor) -> torch.Tensor:
-    return nn.functional.cross_entropy(logits, torch.arange(len(logits), device=logits.device))
+def contrastive_loss(logits: torch.Tensor, dim: int) -> torch.Tensor:
+    neg_ce = torch.diag(nn.functional.log_softmax(logits, dim=dim))
+    return -neg_ce.mean()
```
Comment (sgugger) [inline src/transformers/models/clip/modeling_clip.py:67]: Ah didn't catch that change. Can we focus the change on the TensorFlow model only since the issue was with this one?

Comment (raghavanone) [reply src/transformers/models/clip/modeling_clip.py:67]: @sgugger  This issue affects all the models which uses contrastive loss. Currently none of the models using contrastive loss with return loss parameter works.

Comment (sgugger) [reply src/transformers/models/clip/modeling_clip.py:67]: Please share a snippet of code with the problem as we have multiple users training CLIP without any issue.
